Create basic Go disassembler

bbarenblat opened the pull request:
Create a tool, accessible through ‘go tool disas’, to disassemble Go object files.  The tool is extremely simple – it merely walks through all the sections in the Go object file and runs each section through objdump -D.

Currently, the objdump command lines are hard-coded, so the tool supports only arm and riscv.

bbarenblat committed:
Create basic Go disassembler

Create a tool, accessible through ‘go tool disas’, to disassemble Go
object files.  The tool is extremely simple – it merely walks through
all the sections in the Go object file and runs each section through
objdump -D.

Currently, the objdump command lines are hard-coded, so the tool
supports only arm and riscv.

prattmic commented on src/cmd/disas/main.go:
```diff
@@ -0,0 +1,115 @@
+	return nil
+}
+
+func disas1(sym *goobj.Sym, data []byte) {
```
`disassembleSym`?

mgyenik replied:
If these wrote to an `io.Writer` and had the temp file creation one level higher, it would be easy to have this support disassembling to files, stdout/stderr, etc...

bbarenblat replied:
@prattmic: To what degree do you think we should try to be consistent with style elsewhere in the toolchain? The rest of the repository tends to use terse, non-camel-case names, so I’m hesitant to break from that convention.

bbarenblat replied:
@mgyenik: I don’t fully understand. Currently, `disas1` creates a temporary file to run objdump on and writes results to standard output. I could easily make it write to an `io.Writer`, but that’s orthogonal to the temp file. Are you suggesting that I write results to an `io.Writer` or that I reuse a temporary file for all the disassembly operations (or both)?

prattmic replied:
The short names you have are fine, but could you add an comment explaining the purpose of each?

prattmic commented on src/cmd/disas/main.go:
```diff
@@ -0,0 +1,115 @@
+	}
+}
+
+func disas(file string, pkgpath string) {
```
`disassembleFile`?

bbarenblat replied:
See [comment on `disas1`](#discussion_r48179772).

prattmic commented on src/cmd/disas/main.go:
```diff
@@ -0,0 +1,115 @@
+	log.SetPrefix("disas: ")
+
+	// Ensure that we actually support this architecture.
+	objdumpcmd("dummy")
```
It would be nice if this returned an error or ok.

bbarenblat replied:
[Done](https://github.com/bbarenblat/go/commit/eb155c9551f191a999340881643c30b90250ae2b).

prattmic commented:
LGTM

I only have nits.

mgyenik commented on src/cmd/disas/main.go:
```diff
@@ -0,0 +1,115 @@
+	"cmd/internal/obj"
+)
+
+func objdumpcmd(fname string) *exec.Cmd {
```
I think this was mentioned below. Perhaps this could return an error:

``` go
func objdumpcmd(fname string) (*exec.Cmd, error) {
switch GOARCH {
    case "arm":
        ...
    case "riscv":
        ...
    default:
        return nil, fmt.Errorf("unsupported architecture %s", GOARCH)
    }
}

And it could be used below:
// Ensure that we actually support this architecture.
if _, err := objdumpcmd(""); err != nil {
    log.Fatalf("%s\n", err)
}
```

Or something along those lines. Alternatively you could factor this out into a map of arch string→cmd and test for existence in the map to determine if an arch is supported instead of calling the function with a dummy parameter.

bbarenblat replied:
[Done](https://github.com/bbarenblat/go/commit/eb155c9551f191a999340881643c30b90250ae2b).

I experimented with factoring this out into a map but eventually decided it was a bad idea. Go lacks map constants, so I would either have to create a variable map or wrap it in a function, at which point I’m doing exactly what I’m doing now, but with more unnecessary abstraction.

mgyenik commented:
LGTM, I also have only nits/suggestions, feel free to ignore anything I said and go ahead and merge when you're happy!

bbarenblat committed:
Correctly order defer directives

defer directives are executed in LIFO order, so order them such that the
temporary file is closed and then removed, not the other way around.

bbarenblat committed:
Make objdumpcmd (now called getobjdumpcmd) return an error on failure

bbarenblat commented:
I’d appreciate a quick look at 45c615b, which is unrelated to previous comments.

prattmic commented:
LGTM.

Once you merge this change, I will push it to our gerrithub repo.
